### PR TITLE
feat: Synthetic-porep

### PR DESCRIPTION
**Pull request by snadrus**
## Related Issues
Synthetic PoRep Sealing Implementation for considerable reduction in disk usage by sealing machines during the network proof epochs. 

## Proposed Changes
This including Rust Proofs, FFI, and Lotus Sealing Scheduler but lacks Network/Protocol code.

## Additional Info
https://docs.google.com/document/d/1Ug4uC89Kkwhn3vcbiHETxToIUQ4zXEPa4QcQMPcvRxI/edit#heading=h.277tjch7w86w

https://www.notion.so/pl-strflt/FIP-0059-Synthetic-PoRep-180d995a027647198f6bffebc3eca932

https://www.notion.so/pl-strflt/Synthetic-PoRep-c15c9005f6284a9b9b3adeb6a3cecff1

https://github.com/filecoin-project/FIPs/discussions/245

## Checklist

Before you mark the PR ready for review, please make sure that:

- [ ] Commits have a clear commit message.
- [ ] PR title is in the form of of `<PR type>: <area>: <change being made>`
  - example: ` fix: mempool: Introduce a cache for valid signatures`
  - `PR type`: fix, feat, build, chore, ci, docs, perf, refactor, revert, style, test
  - `area`, e.g. api, chain, state, market, mempool, multisig, networking, paych, proving, sealing, wallet, deps
- [ ] New features have usage guidelines and / or documentation updates in
  - [ ] [Lotus Documentation](https://lotus.filecoin.io)
  - [ ] [Discussion Tutorials](https://github.com/filecoin-project/lotus/discussions/categories/tutorials)
- [ ] Tests exist for new functionality or change in behavior
- [ ] CI is green

### CODE REVIEW

File: chain/actors/builtin/miner/miner.go [hunk 1]
```diff
@@ -228,17 +228,34 @@ func PreferredSealProofTypeFromWindowPoStType(nver network.Version, proof abi.Re
...
+			return abi.RegisteredSealProof_StackedDrg64GiBV1_1, nil
+		default:
+			return -1, xerrors.Errorf("unrecognized window post type: %d", proof)
+		}
+	}
+
 	switch proof {
 	case abi.RegisteredPoStProof_StackedDrgWindow2KiBV1, abi.RegisteredPoStProof_StackedDrgWindow2KiBV1_1:
-		return abi.RegisteredSealProof_StackedDrg2KiBV1_1, nil
+		return abi.RegisteredSealProof_StackedDrg2KiBV1_1_Feat_SyntheticPoRep, nil
```
Comment (arajasek) [inline chain/actors/builtin/miner/miner.go:250]: Nit: not my favourite name

File: storage/sealer/ffiwrapper/sealer_cgo.go [hunk 2]
```diff
@@ -859,6 +859,33 @@ func (sb *Sealer) SealPreCommit2(ctx context.Context, sector storiface.SectorRef
...
+			[]abi.PieceInfo{{Size: abi.PaddedPieceSize(ssize), PieceCID: unsealedCID}})
+		if err != nil {
+			log.Warn("GenerateSynthProofs() failed: ", err)
+			log.Warnf("num:%d tkt:%v seed:%v sealedCID:%v, unsealedCID:%v", sector.ID.Number, ticket, sealedCID, unsealedCID)
+			return storiface.SectorCids{}, xerrors.Errorf("checking PreCommit failed: %w", err)
+		}
+
+		if err = ffi.ClearLayerData(ssize, paths.Cache); err != nil {
+			log.Warn("failed to GenerateSynthProofs(): ", err)
+			log.Warnf("num:%d tkt:%v seed:%v sealedCID:%v, unsealedCID:%v", sector.ID.Number, ticket, sealedCID, unsealedCID)
```
Comment (arajasek) [inline storage/sealer/ffiwrapper/sealer_cgo.go:880]: Nit: not actually logging the seed here?

**Review by arajasek — COMMENTED** [review]
This generally looks good, but we're really gonna need tests in order to gauge correctness. Given that this is just a draft, I'd consider (temporarily) setting `network.SyntheticVersion` to 20, and writing an itest that generates and submits synthetic proofs. This will not fully pass (it'll get rejected by the actors code), but will at least get some of the code paths exercised.

**Comment by arajasek** [comment]
@snadrus Here are 2 tests that can combine to form a good foundation for a test here:

- [TestPledgeSectors](https://github.com/filecoin-project/lotus/blob/08b22edd89ea8365d8126b004b8ab7999f868a6d/itests/sector_pledge_test.go#L26), a very simple test that sets up a miner on a "devnet" and pledges sectors on it.
  - You'll want to tweak the miner config to use SynPoRep, of course
- [TestWindowPostV1P1NV19](https://github.com/filecoin-project/lotus/blob/4ff1d1dc8dd765d99ff16d51a7606232743d21c5/itests/wdpost_test.go#L349), a recent test that configures the "devnet" used by the test to start at a particular network version, and then tests the PoSt scheduling and generation on it. You can apply something similar to your test, I think.

File: chain/actors/builtin/miner/miner.go (hunk 1 above)
Comment (snadrus) [reply chain/actors/builtin/miner/miner.go:250]: It's a bubble-up of the Rust definition.

**Comment by snadrus** [comment]
It seems like I need to wait until the actors work is done as this work revolves around the sealing pipeline but those recommended test require the network side to act correctly.

**Comment by magik6k** [comment]
Merged into https://github.com/filecoin-project/lotus/pull/11057

**Pull request merged**